src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -228,7 +228,7 @@ private static MethodSymbol GetEntryPoint(CSharpCompilation compilation, PEModul
                 !diagnostics.HasAnyErrors())
             {
                 BoundStatement body = synthesizedEntryPoint.CreateBody(diagnostics);
-                if (body.HasErrors || diagnostics.HasAnyErrors())
+                if (body.HasErrors || diagnostics.HasAnyErrors() || !synthesizedEntryPoint.HasBody)
```

Comment:
As @AlekseyTs pointed out above, some of the bodies we generate are not to be emitted. If you take a look in the method `SerializeMethodBodies`, you'll see that it iterates over all method definitions and emits only those for which `MethodDefinition.HasBody` returns true. We're making unnecessary calls to `GenerateMethodBody` and this is making us reach unreachable code (`SynthesizedDelegateSymbol.AreLocalsZeroed`, for instance). I simply copied the HasBody behavior from `MethodDefinition` to `MethodSymbol` in order to avoid unnecessary calls.